service/raft/raft_group0.cc

```diff
@@ -618,6 +769,618 @@ persistent_discovery::persistent_discovery(raft::server_address self, const peer
+    std::unordered_map<gms::inet_address, utils::UUID> versions;
+    while (true) {
+        // We fetch the config on each iteration; some nodes may leave.
+        auto group0_config = group0_server.get_configuration();
+        auto current_config = get_raft_members_inet_addrs(group0_config.current);
+
+        bool failed = false;
+        co_await max_concurrent_for_each(current_config, max_concurrency, [&] (const gms::inet_address& node) -> future<> {
+            if (versions.contains(node)) {
+                // This node was already contacted in a previous iteration.
+                co_return;
+            }
+
+            try {
+                upgrade_log.info("synchronize_schema: `send_schema_check({})`", node);
+                versions.emplace(node,
+                    co_await with_timeout(as, rpc_timeout, [&ms, addr = netw::msg_addr(node)] (abort_source& as) mutable {
+                            return ms.send_schema_check(std::move(addr), as);
+                        }));
+            } catch (abort_requested_exception&) {
+                upgrade_log.warn("synchronize_schema: abort requested during `send_schema_check({})`", node);
+                throw;
+            } catch (...) {
+                // XXX: are there possible fatal errors which should cause us to abort the entire procedure?
+                upgrade_log.warn("synchronize_schema: `send_schema_check({})` failed: {}", node, std::current_exception());
+                failed = true;
+            }
+        });
+
+        if (failed) {
+            upgrade_log.warn("synchronize_schema: there were some failures when collecting remote schema versions.");
+        } else if (group0_config.is_joint()) {
+            upgrade_log.warn("synchronize_schema: group 0 configuration is joint: {}.", group0_config);
+        } else {
+            co_return versions;
+        }
+
+        upgrade_log.info("synchronize_schema: checking if we can finish early before retrying...");
+
+        if (co_await can_finish_early()) {
+            co_return std::nullopt;
+        }
+
+        upgrade_log.info(
+                "synchronize_schema: could not finish early."
+                " Sleeping for a while before retrying remote schema version collection...");
+        co_await sleep_abortable(retry_period, as);
+        if (retry_period < max_retry_period) {
+            retry_period *= 2;
+        }
+    }
+}
+
+// Returning `true` means we synchronized schema.
+// `false` means we finished early after calling `can_finish_early`.
+//
+// Postcondition for synchronizing schema (i.e. we return `true`):
+// Let T0 be the point in time when this function starts.
+// There is a schema version X and a point in time T > T0 such that:
+//     - the local schema version at T was X,
+//     - for every member of group 0 configuration there was a point in time T'
+//       such that T > T' > T0 and the schema version at this member at T' was X.
+//
```

Comment:
I think this definition is too restrictive for eventually consistent schema propagation. After this returns, every replica could have a different schema version for all T' < now(), and yet the local schema can contain all the past schema changes.

What we really want is that local node reflects all schema changes initiated earlier. 
 
Does synchronize_schema() have the same purpose as migration_manager::sync_schema()? Maybe we can merge them into one thing, the less specialized paths we have during upgrade the more reliable it will be.

Reply:
It has the same effect as `sync_schema` when it finishes (for `nodes` = group 0 members), however:
- `sync_schema` uses the `_schema_pulls` queue. If there's a schema pull that is stuck, we're stuck. We may timeout on such a stuck schema pull, but then when we retry, our request will just get stuck again waiting for the existing schema pull that is stuck to finish. I want to be able to timeout, cancel the existing RPC, and send a new RPC (or not - if the unavailable remote got removed from group 0 by the next iteration). Indeed, observe that I'm using `with_timeout` which cancels the RPCs, and then send new RPCs in the next iteration.
- lot's of logging here that happens once during upgrade; we probably don't want this logging during every schema sync
- collecting schema versions step has its own retry loop here
- we check if we can finish early

Reply:
> It has the same effect as sync_schema when it finishes (for nodes = group 0 members)

Actually no, it's stronger.
`synchronize_schema` only leaves after we verified that versions are the same.
`migration_manager::sync_schema` doesn't verify that, only ensures that our versions are more up to date.

Reply:
> I think this definition is too restrictive for eventually consistent schema propagation. After this returns, every replica could have a different schema version for all T' < now(),

Not if you remember the context in which this is executed.
We first ensure everyone is in `synchronize` so nobody is performing any more schema changes (the dry run is an exception but we only use it to improve liveness, not correctness).
Then we call `synchronize_schema`. After it finishes, then we know schema is the same on everyone - because we started `synchronize_schema` after ensuring that nobody makes any more schema changes.
We don't want "eventually consistent schema propagation", we want schema to be completely in sync before we start doing group 0 ops.

Reply:
@tgrabiec I want to guarantee the following during the upgrade procedure:

There is a moment in time `T` such that
- all nodes have the same schema at `T` - call this state `S`
- all schema changes after `T` are serialized through Raft group 0, starting from state `S`.

That's why we:
1. lock schema changes across the entire cluster
2. ensure that schema version is the same across the entire cluster - something that `migration_manager::sync_schema` does not provide (it does a pre-check like that to decide if schema pull is not necessary, but it doesn't do a post-check).
3. only then enter `use_post_raft_procedures` where group 0 operations are available.

Reply:
I understand the motivation now. But notice that if we implement my suggestion to implement upgrade as a group0 command, then we don't have to have as strong guarantees. It's enough that the leader pulls, other nodes can be behind. And now I can use the liveness argument against you, and say that the system is more tolerant to pulls on other nodes being blocked for some reason.

Reply:
> But notice that if we implement my suggestion to implement upgrade as a group0 command, then we don't have to have as strong guarantees. It's enough that the leader pulls, other nodes can be behind. And now I can use the liveness argument against you, and say that the system is more tolerant to pulls on other nodes being blocked for some reason.

But any node can become a leader. What happens when a leader changes from one that could pull to one that cannot?

I think the entire schema business is much simpler to reason about if we know that at some point everybody was in sync, and from that point on everything is nicely serialized. Is there much value in relaxing those guarantees?

Also, it's not clear how complex the command-based solution would be. We would now need a more complex protection from running other commands. Instead of depending on local `group0_upgrade_state`, we would have to read the replicated group 0 state when we perform a schema change (to read this `upgraded` flag). And it's hard to understand how this interacts with recovery. We cannot perform recovery by setting the `upgraded` flag back to `false` - indeed, recovery is for scenarios where group 0 is unavailable.

I propose we don't try redesigning the entire solution from scratch at this point. There was a time for it, I wrote the design, everyone could comment and propose other solutions before this code was written.